### PR TITLE
Converged source traits onto consistent naming, structure and API shape.

### PR DESCRIPTION
**Pull request by AlexSkrypnyk**
# Source convergence pass

Drove `src/` toward internal consistency: one canonical form per concept across naming, structure and imports, API shape, and architecture, then converged source comments onto the technical register. Every change is behavior-preserving. No step text, method name, property name or Behat tag changed, so there is nothing for a consumer to update.

`STEPS.md` is byte-identical before and after.

### Naming (`2d66160`)

67 local-variable renames across 12 files. Body locals only - no signature, member or step text touched.

- Locals named after the wrong entity: media and ECK table rows named `$node_hash`, a page object named `$element`, `findAll()` results named `$nodes` beside siblings calling them `$elements`.
- Abbreviations that already had a spelled-out twin in the same file or a sibling trait: `$doc`, `$ch`, `$attr_*`, `$item`, `$el`, `$idx`, `$cookies_objs`, `$saved`.
- Entity storage handlers unified on `$storage`; `$controller` was Drupal 7/8-era terminology and split the population 3-3.
- Database connection unified on `$database`.

### Structure and imports (`7124921`, `0231115`)

- Import blocks alphabetized in 25 files. Roughly half the traits hoisted the `Behat\Step\*` imports above everything else and several were internally unsorted; the block is now a single flat list sorted case-insensitively by FQCN. Applied by script, so it is idempotent.
- Generic array annotations always carry a space after the comma.
- 16 doubled blank comment lines collapsed to a single separator, plus the one stray blank line after a method's opening brace.
- Three `@throws \Exception` tags replaced with the exception types the methods actually throw.

### API shape (`1bb9400`)

- The single `assert()` type check became an instanceof guard that throws, and the single `private` helper became `protected` - both required by the project rules.
- `strval()`/`intval()` replaced with cast operators, one double cast collapsed, and the two `waitFor` union types aligned.
- `is_null()` replaced with strict `=== NULL`, eliminating the third of three null-checking forms.
- Container access uses the dedicated `\Drupal` facade where one exists, never `getContainer()->get()`.
- Concatenated exception messages use `sprintf()`, with byte-identical output.
- Instance helpers reached through `self::`/`static::` are now called on `$this`.
- `PyStringNode` bodies convert through `getRaw()`; JS-embedded JSON uses `JSON_UNESCAPED_SLASHES` everywhere.

### Architecture (`b652254`)

The seven-line block that fetches the current URL, extracts its query string and parses it was copied verbatim into all four `PathTrait` query-parameter assertions. It now lives in `pathGetCurrentUrlQuery()`. All 27 `path.feature` scenarios pass unchanged.

### Comment register (`0fd241a`)

Comment-only diff: 27 files, 85 insertions, 145 deletions. Comments that only restated the statement below them were deleted; the rest were reworded to state what the code does rather than narrate it.

Scope was inline comments and the prose of protected-helper and property docblocks. Trait docblocks, step docblocks and their `@code` examples were deliberately left alone - they are the published documentation that `docs.php` renders into `STEPS.md`.

**This pass is incomplete.** It reached 27 of 51 files before the run was cut short. Not processed: `FieldTrait`, `TableTrait`, `KeyboardTrait`, `LinkTrait`, `CommandTrait`, `JavascriptTrait`, `DiagnosticsTrait`, the generic `HelperTrait`, `WaitTrait`. Processed only in part: `EmailTrait`, `UserTrait`, `ConfigTrait`, `MetatagTrait`, `FileDownloadTrait`, `CookieTrait`, `PathTrait`, `DateTrait`, `ResponsiveTrait`. Re-running the pass over those files is safe and idempotent.

One constraint worth recording, hit independently by every reviewer: the Drupal phpcs standard mandates a docblock on every method and property, so a summary that merely restates the member name cannot be deleted even though it earns nothing. Those were treated as contract - register-checked, never removed.

## Comments found factually wrong or stale

None of these were fixed. A comment that misstates what the code does is a behavior-documentation defect and deserves its own reviewed change, not a line buried in a wording diff.

| Location | Claim | Reality |
|---|---|---|
| `ModalTrait.php:231` | "Find the first visible modal, or fall back to the first DOM match" | The fallback is per selector, not global. The first selector in `modalGetSelectors()` with any match wins, so an invisible `.ui-dialog` is returned even when a later selector matches a genuinely visible modal. This one looks like a real defect, not just a wrong comment. |
| `ContentBlockTrait.php:168` | Entities are stored in "the static `$blockContentEntities` array" | No such property exists. Cleanup goes through `entityRegister()`. |
| `MediaTrait.php:347` | `@param` "The node type", `@return` "Array of node ids" | The method queries media entities and returns media IDs. |
| `MediaTrait.php:309` | "re-use of the functionality provided by DrupalExtension" | `expandEntityFields()` lives in `drupal/drupal-driver`, not DrupalExtension. |
| `UserTrait.php:442` | `userLoadByName()` returns `?UserInterface`, NULL when missing | It throws instead; the nullable branch is dead. |
| `MenuTrait.php:15` | Trait "asserts menu items" and "menu link visibility" | The trait defines only `Given` steps; no assertions exist. |
| `BlockTrait.php:353` | `@throws \Exception` when no block is found | The method never throws; it returns NULL. |
| `WatchdogTrait.php:182` | "all logged entries for PHP channel" | Filters on every configured type, not only `php`. |
| `WatchdogTrait.php:169` | "no errors above the severity threshold" | Entries at the threshold fail too; the `@throws` line says "at or above". |
| `TestmodeTrait.php:40` | "Disable test mode before test run" | It is an `AfterScenario` hook. |
| `AccessibilityTrait.php:66` | JS cache "fetched once per run" | The cache is static, so per process; a parallel run spans several. |
| `AccessibilityTrait.php:527` | "see class docblock for the full structure" | The trait docblock never documents that structure. |
| `ElementTrait.php:1177` | JS copied from `tests/behat/fixtures/relative.html` | That file does not exist; the fixture is `elements_relative.html`. |
| `ElementTrait.php:1177` | `elementIsVisuallyVisible()` returns `bool` | No native return type; returns the raw `mixed` from `evaluateScript()`. |
| `ElementTrait.php:1171` | `@param` "(optional) ... Defaults to 0" | The parameter has no default. |
| `KeyboardTrait.php:86` | `$selector` behavior applies "if omitted" | The parameter has no default and cannot be omitted. |
| `KeyboardTrait.php:132` | Exception text points at `keyPress($char)` | No such method exists. |
| `FieldTrait.php:409` | `fieldGetRequiredMarkerSelectors()` is the override point used by `fieldIsMarkedRequired()` | It is never called; the checks are hardcoded. |
| `XmlTrait.php:762` | "Skip the default XML namespace" | Skips the reserved `xml:` namespace; a default namespace is an unprefixed `xmlns=`. |
| `DateTrait.php:82` | `@code` example begins `Give content ...` | Malformed Gherkin keyword; the example would not parse. |
| `CookieTrait.php:95` | `@code` shows `a cookie with name "old_session"` | The registered pattern requires `the name`; the example would not match. |
| `PathTrait.php:165` | Summary duplicated from the no-parameter variant | It misdescribes one of the two steps. |
| `XmlTrait.php:858` | `xmlValidateDtd()` shares one code path "without exposing external-entity loading" | Verified on this project's runtime (PHP 8.3.30, libxml 2.13.9): under plain `LIBXML_DTDVALID` a `SYSTEM` entity **is** dereferenced - a nonexistent path yields `failed to load "file:///..."`. The entity content is *not* substituted into the document, because `LIBXML_NOENT` is not passed, so there is no read-and-exfiltrate path; but "without exposing external-entity loading" overstates the guarantee. Pre-existing on `main`; the comment pass restructured the sentence without touching the claim, exactly as its rules require. |

## Judgment calls deferred to you

Full-auto applies confident, behavior-preserving convergences and leaves the rest to a human. These are real findings with evidence, not oversights.

### Public surface - every one of these is a breaking change for consumers

These traits are mixed into consumer `FeatureContext` classes, so step text, method names, property names and tag names are all public API.

- **Step-text rule violations.** `Given I accept all confirmation dialogs` and its negative violate the no-`Given I` rule; `I should (not) see the modal` violates Then-entity-first; `:number` is used for a pixel offset where siblings use `:tolerance`; 11 steps carry optional `(s)` tokens.
- **Step-text drift.** Placeholder names for one concept (`:attribute`/`:attribute_name`, `:text`/`:value`, `:vocabulary`/`:vocabulary_machine_name`/`:machine_name`, `:content_type`/`:type`, `:media_type`/`:bundle`, `:locator`/`:selector`) across ~30 steps; article and label drift across ~15; ~15 `Given` steps that use `is`/`has been` instead of `exists`/`have`.
- **Method-name rule violations.** `MenuTrait`'s `loadMenuByLabel()`/`loadMenuLinkByTitle()` carry no trait prefix; Drupal `HelperTrait` uses an `entity*` family rather than `helper*`; `draggableViews*` capitalises the V against its own trait name; `waitWaitForSeconds` doubles the prefix.
- **Naming drift.** `Normalize` (4) vs `Normalise` (2); six coexisting negation shapes (`NotExists`, `IsNot`, `HasNo`, `AssertNo`, antonyms) across ~25 methods; `Get*` getters vs bare-noun getters on 11 documented override points.
- **Visibility and shape.** Three public non-step callables; `DateTrait`'s all-static paradigm against instance helpers everywhere else; the unprefixed `DEFAULT_WAIT_TIMEOUT` constant; `FieldTrait` composing `KeyboardTrait` without ever calling it (removing it would strip keyboard steps from consumers).
- **Tag mechanics.** `behat-steps-skip:` uses `__FUNCTION__` tokens in 13 places and PascalCase trait names in 4; eight hooks support no skip tag at all.

### Behavior-changing

- **Assertion exception type.** `ExpectationException` dominates (~200 sites) but `\Exception` is used in ~50, plus `\RuntimeException` leaking from `xmlAssertResponseIsXml` where the JSON twin throws `ExpectationException`. Structurally blocked in places: `CommandTrait` and `ConfigTrait` have no Mink session, so the driver argument is unavailable.
- **Duplicate-entity tie-break.** `blockLoadByLabel()` uses `krsort()` + `end()`, returning the *oldest* match, while every sibling uses `ksort()` + `end()` for the newest. Likely a latent bug.
- **Comparison strictness.** Five `in_array()` calls without the strict flag, six loose `==`/`!=` scalar comparisons.
- **Missing contrib-module guards.** `MenuTrait` imports `menu_link_content` and `DraggableviewsTrait` writes a contrib table with no availability guard; `SearchApiTrait` calls `search_api_entity_insert()` unguarded.
- **Guard outcome.** Silent-return versus throw for the same class of failure, across JS-driver gating, `EmailTrait`'s `StatementInterface` fallback, and `LinkTrait`'s negative step passing when the container is missing.
- **Exception message trailing period.** The plan sized this at 16 sites; it is actually about 50 of roughly 400 messages. The omissions are not random - `CommandTrait`, `PathTrait`, `JsonTrait`, `XmlTrait`, `ModuleTrait` and `EmailTrait` consistently drop the period when the message ends in an interpolated value (`: %s`, `"%s"`). Several are asserted verbatim by in-repo BDD features. At that size and with that coupling it stops being a surgical wording fix, so it is left for a deliberate decision.
- **`DateTrait` regex and capture.** Both failure branches interpolate `$matches[1]` (the token keyword) rather than the offending value, and the token match uses a `[relative:]+` character class instead of a literal.
- **`emailFollowLinkNumber()` accepts an out-of-range link number.** Confirmed by walking the code: with a link number of `0`, the bound check `count($links) < $link_number` is false, so execution reaches `$links[-1]`, which is undefined on a zero-indexed list. PHP 8 emits `Undefined array key -1` and evaluates to `NULL`, which then reaches `Session::visit()`. Negative numbers behave the same. Pre-existing - this branch only changed `intval($link_number)` to `(int) $link_number`, which is exactly equivalent - and it applies to `emailFollowLinkNumberWithSubjectContaining()` too. Fixing it needs a guard plus a test for the `0` and negative cases.

### Genuine ties - no canonical form to pick

- Truthiness `!$x` (~40) versus strict `=== NULL` (~35) for nullable objects, with opposing per-trait majorities.
- `empty($array)` (~30) versus `$array === []` (~25), with the newest code leaning strict.
- `self::` versus `static::` for static helpers (13 each); switching to `static::` would widen consumer override semantics and should be a deliberate design decision.
- `@param`-before-`@code` (7) versus `@code`-before-`@param` (6), and flush versus indented `@code` examples.

### Left to a dedicated pass

- Member ordering within traits: the dominant layout is clear, but reordering reshuffles `STEPS.md` output for zero runtime gain.
- Remaining duplication: verbatim twins across traits (`xmlReadFile`/`jsonReadFile`, the JS error-buffer read, the module-exists check), duplicates that have drifted behaviorally (two cURL implementations with different timeouts and status handling, two cookie driver-sniffing cascades), and ~35 in-file twin bodies.
- `@codeCoverageIgnore` marker drift: equivalent defensive guards are marked in some traits and not others. `codecov/patch` and `codecov/project` are required checks, so unifying needs a coverage re-baseline rather than a mechanical sweep.

## Verification

- **Full BDD suite: 1036 scenarios, 1036 passed; 4376 steps, 4376 passed.** Browser scenarios included - 197 `@javascript` scenarios ran, `modal.feature` and `dropzone.feature` among them.
- `ahoy lint` (phpcs, phpstan level 7, rector dry-run, gherkinlint): green after every class.
- `ahoy test-unit`: 392 tests, 677 assertions, green after every class.
- `ahoy lint-docs`: `STEPS.md` unchanged throughout.
- `path.feature` run on its own after the `PathTrait` extraction: 27 scenarios, 105 steps, green.

Two earlier full-suite attempts aborted with Chrome `tab crashed` - at `modal.feature:9` once and at `cookie.feature:84` (a vendor `I visit` step) the next time. The crash moved between runs over identical code and both features passed in isolation, so it is a browser resource flake; the clean run above confirms it. Worth knowing that Behat throws out of the whole run when this happens, so every feature after the crash point silently goes unexecuted while the exit code reads as a normal failure.

Also worth knowing for future narrowing: passing `--tags` on the command line *replaces* the `~@skipped` filter that `behat.yml` sets, rather than intersecting with it, so `@skipped` scenarios start running and failing. Compose it as `~@skipped&&<your tag>`.

On CI, one matrix job (`PHP 8.3, Drupal 10, Deps normal`) failed on the first run at 1m34s while its twelve siblings each took 10-16 minutes. The log showed `502 Bad Gateway` from `registry-1.docker.io` while pulling a `uselagoon/php-8.3-cli-drupal` layer, so it never reached the test suite. A single re-trigger cleared it and that job now passes in 13m41s.

## Before / After

The clearest single instance of "many copies, one canonical form" is `PathTrait`'s query-parameter assertions, where an extracted helper replaced four verbatim copies of the same parsing block. The same collapse-to-one-form pattern repeats without extraction across the naming, import-ordering and API-shape convergences elsewhere in `src/`.

```
BEFORE - PathTrait
┌────────────────────────────────────────────────────────────────┐
│ pathAssertUrlHasParameter()              ┐                     │
│ pathAssertUrlHasParameterWithValue()     │  each repeats the   │
│ pathAssertUrlHasNoParameter()            │  same 7 lines       │
│ pathAssertUrlHasNoParameterWithValue()   ┘                     │
│                                                                │
│     $url = $this->getSession()->getCurrentUrl();               │
│     $url_query = parse_url((string) $url, PHP_URL_QUERY);      │
│     $url_query = $url_query === FALSE ? '' : (string) ...;     │
│     $q = [];                                                   │
│     parse_str($url_query, $q);                                 │
│                                                                │
│ 4 copies, byte-identical, free to drift apart unnoticed        │
└────────────────────────────────────────────────────────────────┘

AFTER - PathTrait
┌────────────────────────────────────────────────────────────────┐
│ pathGetCurrentUrlQuery()   the 7 lines, defined exactly once   │
│                                                                │
│ pathAssertUrlHasParameter()              ┐                     │
│ pathAssertUrlHasParameterWithValue()     │  $query = $this->   │
│ pathAssertUrlHasNoParameter()            │  pathGetCurrentUrl  │
│ pathAssertUrlHasNoParameterWithValue()   ┘  Query();           │
│                                                                │
│ 1 helper + 4 one-line call sites, one place to fix or extend   │
└────────────────────────────────────────────────────────────────┘
```

### CODE REVIEW

File: src/Drupal/EmailTrait.php [hunk 1]
```diff
@@ -402,7 +392,7 @@ public function emailAssertMessageFieldNotExact(string $field, PyStringNode $str
    */
   #[When('I follow link number :link_number in the email with the subject :subject')]
   public function emailFollowLinkNumber(string $link_number, string $subject): void {
-    $link_number = intval($link_number);
+    $link_number = (int) $link_number;
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/Drupal/EmailTrait.php:395]: Confirmed as a real defect, but not adopted in this PR.

Verified the failure mode: with a link number of `0`, the bound check `count($links) < $link_number` is false, so execution reaches `$links[$link_number - 1]`, which is `$links[-1]`. `emailExtractLinks()` returns a zero-indexed list, so that key is undefined: PHP 8 emits `Undefined array key -1` and evaluates to `NULL`, which then reaches `Session::visit()`. A negative number behaves the same way.

Two reasons it is not being fixed here:

1. It predates this branch. The only change to that line is `intval($link_number)` becoming `(int) $link_number`, which is exactly equivalent for every input, so the diff neither introduces nor widens the defect. The line appears in the diff purely because of the cast-style convergence.

2. This PR is a behavior-preserving convergence pass, and its description commits to that explicitly. Adding a guard changes the failure mode for input the library currently accepts, so it belongs in its own change where it can be reviewed as a behavior fix and paired with a test for the `0` and negative cases.

Recorded in the PR description under the deferred behavior-changing findings so it is not lost.

File: src/XmlTrait.php [hunk 2]
```diff
@@ -861,13 +855,13 @@ protected function xmlValidateRelaxNg(string $schema): void {
...
-   * path without exposing external-entity loading.
+   * The DTD is embedded as an internal subset and the response is reloaded
+   * with validation enabled. A DTD from a file and an inline DTD share this
+   * code path without exposing external-entity loading.
    *
    * DTDs are namespace-unaware, so a namespaced response is validated verbatim
-   * and its `xmlns` attributes must be declared in the DTD, matching native DTD
-   * validation semantics.
+   * and its `xmlns` attributes must be declared in the DTD. This matches
+   * native DTD validation semantics.
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/XmlTrait.php:864]: Half right, and worth pinning down precisely. I probed the actual behaviour on this project's runtime (PHP 8.3.30, libxml 2.13.9), reproducing the exact load `xmlValidateDtd()` performs.

**External entities are dereferenced.** A `SYSTEM` entity pointing at a nonexistent path produces `failed to load "file:///nonexistent/path/probe.txt": No such file or directory` under plain `LIBXML_DTDVALID`. So the docblock's "without exposing external-entity loading" is inaccurate as written, and you are right to flag it.

**The content is not substituted.** With a `SYSTEM` entity pointing at a real file containing a sentinel value, the document loads with zero errors and the sentinel does not appear anywhere in the serialised output. Substitution requires `LIBXML_NOENT`, which this code does not pass. So the classic XXE read-and-exfiltrate path is not open here; what remains is narrower, essentially a probe channel, since whether the fetch failed is observable through the libxml errors that `xmlFormatErrors()` folds into the assertion message.

Not changed in this PR, for two reasons:

1. Both the flag and the claim predate this branch. The diff only restructures the sentence: the comment pass that touched it rewrites wording and is explicitly forbidden from altering a claim, which is why the inaccurate wording survived rather than being quietly corrected.

2. Passing an additional flag changes which DTDs validate, and this PR is a behavior-preserving pass. A security-relevant parser change deserves its own PR and its own test rather than being buried in a convergence diff.

Worth weighing when that change is made: the DTD is authored by the test writer in their own `Then the response should match the following DTD:` step, not supplied by an untrusted remote party, so the threat model is weaker than for a general-purpose XML parser.

The inaccurate docblock claim is recorded in the PR description's list of comments found factually wrong, which is where that pass reports rather than fixes them.
